### PR TITLE
Move and improve lastSamplePair. 

### PR DESCRIPTION
**Pull request by beorn7**
@juliusv @fabxc This will help federation a lot, and thanks to the shortcut, it will also improve most instant evaluations.

### CODE REVIEW

**Comment by beorn7** [comment]
In combination with #1405 , this change has lead to a 3x performance improvement for rule evaluation on a real-world Prometheus server at Soundcloud.

@mwitkow This will probably also help in your case. It's a completely different approach from the one described in #1264 but it solves the other half of the same problem. With the speed improvements I have seen now in real life, I suspect this actually fixes the more important part, and the index lookups are not that bad, given the cache size is tuned up sufficiently.

@juliusv Do you see any chance reveviewing this and #1405 any time soon?

This is built on top of #1405, and I'm already working on the next improvement built on top of this change... so I'm three levels ahead by now...

**Comment by mwitkow** [comment]
Awesome! Please let me know when there's a version we can try out :)

**Comment by beorn7** [comment]
@mwitkow You could just check out this branch, build, and try it out (but better not in production... ;).

**Comment by beorn7** [comment]
Merged the new commits from beorn7/storage.

@juliusv I assume this is easier to review...

**Comment by beorn7** [comment]
@mwitkow Too happy too soon. Extrapolation is hard. With enough historical churn accumulated, the rule evaluation in my test setup is again dominated by index lookup.

This PR is still a huge improvement. Just doesn't help #1264 as much as initially thought.

@juliusv As said, this is the middle piece between two PRs you have already reviewed. Should be relatively light weight.

**Comment by beorn7** [comment]
@fabxc May I ask you to review this? @juliusv has better context (he reviewed the change this is based on, _and_ the change that depends on this, but this PR is now sitting waiting for 12 days, and @juliusv is on vacation right now).

File: storage/local/interface.go [hunk 1]
```diff
@@ -91,6 +94,17 @@ type Preloader interface {
...
+		fp model.Fingerprint,
+		timestamp model.Time, stalenessDelta time.Duration,
+	) (SeriesIterator, error)
 	// Close unpins any previously requested series data from memory.
 	Close()
 }
+
+// ZeroSamplePair is the pseudo zero-value of model.SamplePair used by the local
+// package to signal a non-existing sample. It is a SamplePair with timestamp
+// model.Earliest and value 0.0. Note that the natural zero value of SamplePair
```
Comment (fabxc) [inline storage/local/interface.go]: The second half does not belong into the public documentation.

File: storage/local/series.go [hunk 2]
```diff
@@ -377,9 +386,27 @@ func (s *memorySeries) newIterator(pinnedChunkDescs []*chunkDesc) SeriesIterator
...
 	from model.Time, through model.Time,
-	fp model.Fingerprint, mss *memorySeriesStorage,
+	lastSampleOnly bool,
+	mss *memorySeriesStorage,
 ) ([]*chunkDesc, SeriesIterator, error) {
+	// If we have to preload for only one sample, and we have a
+	// lastSamplePair in the series, and thas last samplePair is in the
+	// interval, just take it in a singleSampleSeriesIterator. No need to
+	// pin or load anything.
+	if lastSampleOnly {
```
Comment (fabxc) [inline storage/local/series.go]: This shouldn't be a flag but a separate `preloadChunksForInstant` method.

File: storage/local/interface.go (hunk 1 above)
Comment (beorn7) [inline storage/local/interface.go]: You mean the part about the "natural zero value"? I find that quite valuable to answer the FAQ that will inevitably show up for anybody who looks at this variable.

File: storage/local/series.go (hunk 2 above)
Comment (beorn7) [inline storage/local/series.go]: In principle yes, but it's just pass-through from preloadChunksForRange in memorySeriesStorage, and reworking this one is a bit more cumbersome.

I'll give it a try anyway, and then you con decide if you like it more or not.

**Comment by beorn7** [comment]
@fabxc Done. Please have a look if you like it more.

File: storage/local/interface.go (hunk 1 above)
Comment (fabxc) [inline storage/local/interface.go]: Since there's not meaningful way to separate it, sure, let's keep it.

**Comment by fabxc** [comment]
Yep, that seems more consistent :+1:

**Pull request merged**